### PR TITLE
Add xml type support for configurable TOML configuration

### LINKED ISSUE #31325: Configurable `table<map>` and `map[]` are not considered as ambiguous for union type
**Description:**
Consider the following code, 
```ballerina
type Foo record {
    int i;
};

configurable Foo[]|table<Foo> a = ?;
```
This can be provided using 
```toml
[[a]]
i = 22
```
As the value can be assigned to both types, it should give an ambiguous error. 
Currently it creates a table value. 

**Affected Versions:**
beta2- SNAPSHOT

### LINKED ISSUE #31738: Support configurable `xml` type variables through Toml files
**Description:**
As per the spec, the xml typed values can be provided as `string` values through TOML files. 
https://github.com/ballerina-platform/ballerina-spec/blob/master/configurable/spec.md

This feature is currently not supported. We need to implement this support. 
 

**Suggested Labels (optional):**
Team/jBallerina , Area/Configurable

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
- Support configuration of xml type configurable variables through TOML syntax
- Fix value ambiguity that can happen for types that have similar TOML representations.
       - `string` & `xml` types
       - `float` & `decimal` types
       - `map<anydata>[]` & `table` types 

Fixes #31738 
Fixes #31325

## Approach
The xml values can be provided through TOML string.

## Samples
```ballerina
configurable xml xmlVar = ?;
```

In `Config.toml`
```toml
xmlVar="<book><name>Sherlock Holmes</name><author>Arthur Conan Doyle</author></book>"
```

For the ambiguity case,
``` ballerina
configurable float|decimal a = ?;
```
The above sample will throw an error
```
[Config.toml:(2:8,2:13)] ambiguous target types found for configurable variable 'a' with type '(float|decimal)'
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigResolver.java [hunk 1]
```diff
@@ -113,6 +113,8 @@ public ConfigResolver(Map<Module, VariableKey[]> configVarMap, RuntimeDiagnostic
             case TypeTags.RECORD_TYPE_TAG:
                 return getConfigValue(key, configProvider -> configProvider
                         .getAsRecordAndMark(module, key));
+            case TypeTags.XML_TEXT_TAG:
+                return getConfigValue(key, configProvider -> configProvider.getAsXmlAndMark(module, key));
```
Comment (pramodya1994) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigResolver.java:117]:
```suggestion
                return getConfigValue(key, configProvider -> configProvider
                        .getAsXmlAndMark(module, key));
```
Shall we use the same format used in the class?

Comment (pramodya1994) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigResolver.java:117]: Also in L138

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/RuntimeUtils.java [hunk 2]
```diff
@@ -103,6 +104,18 @@ public static boolean isByteLiteral(int intValue) {
         return (intValue >= BBYTE_MIN_VALUE && intValue <= BBYTE_MAX_VALUE);
     }
 
+    /**
+     * Provide the readonly Xml Value that is equivalent to a given string value.
+     *
+     * @param value string value
+     * @return immutable Xml value
+     */
+    public static BXml createReadOnlyXmlValue(String value) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/RuntimeUtils.java:113]: Lets move this to ValueUtils

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/ConfigValueCreator.java [hunk 3]
```diff
@@ -116,6 +127,12 @@ private BArray createArrayValue(TomlNode tomlValue, ArrayType arrayType) {
     private BArray getNonPrimitiveArray(TomlNode tomlValue, ArrayType arrayType,
                                         Type elementType) {
         switch (elementType.getTag()) {
+            case TypeTags.XML_ATTRIBUTES_TAG:
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/ConfigValueCreator.java:130]: Lets move this to seperate method

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java [hunk 4]
```diff
@@ -256,8 +284,102 @@ public static Field createAdditionalField(RecordType recordType, String fieldNam
         }
     }
 
+    private static Object validateAndGetTableArrayValue(TomlTableArrayNode tomlNode, Set<TomlNode> visitedNodes,
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: Lets move this to compile time. May be we can do with a seperate PR.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java [hunk 5]
```diff
@@ -185,13 +220,6 @@ static TomlType getEffectiveTomlType(Type expectedType, String variableName) {
                 return TomlType.TABLE;
             case TypeTags.TABLE_TAG:
                 return TomlType.TABLE_ARRAY;
-            case TypeTags.XML_ATTRIBUTES_TAG:
-            case TypeTags.XML_COMMENT_TAG:
-            case TypeTags.XML_ELEMENT_TAG:
-            case TypeTags.XML_PI_TAG:
-            case TypeTags.XML_TAG:
-            case TypeTags.XML_TEXT_TAG:
-                throw new ConfigException(CONFIG_TOML_TYPE_NOT_SUPPORTED, variableName, expectedType.toString());
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:194]: Shall we remove these messages from message.properties and other places.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java (hunk 4 above)
Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: I'll address this in issue https://github.com/ballerina-platform/ballerina-lang/issues/31889

Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: I'll address this in issue https://github.com/ballerina-platform/ballerina-lang/issues/31889

Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: I'll address this in issue https://github.com/ballerina-platform/ballerina-lang/issues/31889

Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: I'll address this in issue https://github.com/ballerina-platform/ballerina-lang/issues/31889

Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: I'll address this in issue https://github.com/ballerina-platform/ballerina-lang/issues/31889

Comment (HindujaB) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/Utils.java:287]: I'll address this in issue https://github.com/ballerina-platform/ballerina-lang/issues/31889

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**